### PR TITLE
drivers: sensors: smart-battery with SBS 1.1 compliant fuel gauge

### PR DESCRIPTION
**Pull request by grillchef**
Signed-off-by: MORGER Patrick <patrick.morger@leica-geosystems.com>

### CODE REVIEW

File: tests/drivers/sensor/smart_battery/src/test_smartbattery.c [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
+#include <drivers/sensor.h>
+
+#if DT_NODE_HAS_STATUS(DT_ALIAS(bat0), okay)
+#define BAT_NODE_ID	DT_ALIAS(bat0)
+#define BAT_DEV_NAME	DT_LABEL(BAT_NODE_ID)
+#else
+#error "Fuel Gauge device not found"
+#endif
+
+static ZTEST_BMEM const struct device *dev;
```
Comment (alexanderwachter) [inline tests/drivers/sensor/smart_battery/src/test_smartbattery.c:18]:
```suggestion
static ZTEST_BMEM const struct device *dev = DEVICE_DT_GET(DT_ALIAS(bat0));
```

File: tests/drivers/sensor/smart_battery/src/test_smartbattery.c [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+
+const struct device *get_fuel_gauge(void)
+{
+	return device_get_binding(BAT_DEV_NAME);
+}
```
Comment (alexanderwachter) [inline tests/drivers/sensor/smart_battery/src/test_smartbattery.c:23]: No need if you use the change above

File: tests/drivers/sensor/smart_battery/src/test_smartbattery.c [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+
+void test_smartbattery_getDesiredChargingCurrent(void)
+{
+	test_smartbattery_getSensorValue(SENSOR_CHAN_GAUGE_DESIRED_CHARGING_CURRENT);
+}
+
+void test_smartbattery_getAccelX(void)
+{
+	test_smartbattery_getSensorValue_NotSupported(SENSOR_CHAN_ACCEL_X);
+}
```
Comment (alexanderwachter) [inline tests/drivers/sensor/smart_battery/src/test_smartbattery.c:135]: Why so many not supported tests?
One would be enough.

**Review by alexanderwachter — CHANGES_REQUESTED** [review]

File: drivers/sensor/smart_battery/smart_battery.c [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+	int status;
+
+	status = i2c_burst_read(smart_battery->i2c, config->i2c_addr, reg_addr,
+				i2c_data, 2);
+	if (status < 0) {
+		LOG_ERR("Unable to read register");
+		return -EIO;
+	}
+
+	*val = (i2c_data[1] << 8) | i2c_data[0];
```
[the remark this answers is not in the document]
Comment (grillchef) [reply drivers/sensor/smart_battery/smart_battery.c:33]: thx for the hint

File: drivers/sensor/smart_battery/smart_battery.c [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+#include <sys/__assert.h>
+#include <string.h>
+#include <sys/byteorder.h>
+
+#include "smart_battery.h"
+
+static int smart_battery_command_reg_read(const struct device *dev, uint8_t reg_addr,
+				    int16_t *val)
+{
+	struct smartbattery_data *smart_battery = dev->data;
```
[the remark this answers is not in the document]
Comment (grillchef) [reply drivers/sensor/smart_battery/smart_battery.c]: macros added

File: drivers/sensor/smart_battery/smart_battery.h [hunk 6]
```diff
@@ -0,0 +1,81 @@
+#ifndef ZEPHYR_DRIVERS_SENSOR_SMART_BATTERY_H_
+#define ZEPHYR_DRIVERS_SENSOR_SMART_BATTERY_H_
+
+#include <stdint.h>
+#ifndef WIN32
+#include <logging/log.h>
+LOG_MODULE_REGISTER(smart_battery, CONFIG_SENSOR_LOG_LEVEL);
```
[the remark this answers is not in the document]
Comment (grillchef) [reply drivers/sensor/smart_battery/smart_battery.h]: moved into the .c file

File: drivers/sensor/smart_battery/smart_battery.c [hunk 7]
```diff
@@ -0,0 +1,268 @@
...
+/**
+ * @brief sensor value get
+ *
+ * @return -ENOTSUP for unsupported channels
+ */
+static int smartbattery_channel_get(const struct device *dev, enum sensor_channel chan,
+			       struct sensor_value *val)
+{
+	struct smartbattery_data *smart_battery = dev->data;
+	float int_temp;
```
[the remark this answers is not in the document]
Comment (grillchef) [reply drivers/sensor/smart_battery/smart_battery.c]: float removed

File: tests/drivers/sensor/smart_battery/src/test_smartbattery.c (hunk 2 above)
Comment (grillchef) [reply tests/drivers/sensor/smart_battery/src/test_smartbattery.c]: function get_fuel_gauge removed

File: drivers/sensor/smart_battery/smart_battery.c [hunk 8]
```diff
@@ -0,0 +1,269 @@
...
+
+	case SENSOR_CHAN_GAUGE_AVG_CURRENT:
+		val->val1 = smart_battery->avg_current;
+		break;
+
+	case SENSOR_CHAN_GAUGE_TEMP:
+		int_temp = (smart_battery->internal_temperature * 0.1);
+		int_temp = int_temp - 273.15;
+		val->val1 = (int32_t)int_temp;
+		val->val2 = (int_temp - (int32_t)int_temp) * 1000000;
```
[the remark this answers is not in the document]
Comment (grillchef) [reply drivers/sensor/smart_battery/smart_battery.c]: changed as proposed

File: drivers/sensor/smart_battery/smart_battery.c [hunk 9]
```diff
@@ -0,0 +1,269 @@
...
+		break;
+
+	case SENSOR_CHAN_GAUGE_STATE_OF_CHARGE:
+		val->val1 = smart_battery->state_of_charge;
```
[the remark this answers is not in the document]
Comment (grillchef) [reply drivers/sensor/smart_battery/smart_battery.c:77]: is in percentage

File: drivers/sensor/smart_battery/smart_battery.c [hunk 10]
```diff
@@ -0,0 +1,275 @@
...
+		break;
+
+	case SENSOR_CHAN_GAUGE_FULL_CHARGE_CAPACITY:
```
[the remark this answers is not in the document]
Comment (grillchef) [reply drivers/sensor/smart_battery/smart_battery.c:80]: Agree, mAh is not SI.

File: drivers/sensor/smart_battery/smart_battery.c [hunk 11]
```diff
@@ -0,0 +1,275 @@
...
+
+	case SENSOR_CHAN_GAUGE_FULL_AVAIL_CAPACITY:
+		val->val1 = smart_battery->full_avail_capacity;
+		break;
+
+	case SENSOR_CHAN_GAUGE_TIME_TO_EMPTY:
+		val->val1 = smart_battery->time_to_empty;
+		break;
+
+	case SENSOR_CHAN_GAUGE_TIME_TO_FULL:
```
[the remark this answers is not in the document]
Comment (grillchef) [reply drivers/sensor/smart_battery/smart_battery.c:100]: time is in minutes

File: drivers/sensor/smart_battery/smart_battery.c [hunk 12]
```diff
@@ -0,0 +1,269 @@
...
+			       struct sensor_value *val)
+{
+	struct smartbattery_data *smart_battery = dev->data;
+	float int_temp;
+
+	val->val2 = 0;
+
+	switch (chan) {
+	case SENSOR_CHAN_GAUGE_VOLTAGE:
+		val->val1 = smart_battery->voltage;
```
[the remark this answers is not in the document]
Comment (grillchef) [reply drivers/sensor/smart_battery/smart_battery.c]: changed

File: tests/drivers/sensor/smart_battery/src/test_smartbattery.c (hunk 3 above)
Comment (grillchef) [reply tests/drivers/sensor/smart_battery/src/test_smartbattery.c:124]: for each value of the enum a test. Should we use one function with several ranges?

Comment (alexanderwachter) [reply tests/drivers/sensor/smart_battery/src/test_smartbattery.c:124]: I would test only one. The enum will definitely be exchanged in the future and therefor you anyway can't check all.

**Pull request merged**